File: server/downloader/factory.py
from enum import Enum
from typing import Literal, Optional
from urllib.parse import urlsplit

from .AbsDownloader import AbsDownloader
from .hoyolab import HoyolabDownloader
from .twitter import TwitterDownloader


class PageType(Enum):
    twitter = "twitter"
    hoyolab = "hoyolab"


class Scrapper(AbsDownloader):
    def getInstance(
        self,
        url: str = None,
        type: Optional[PageType] = None,
        type_name: str | None = None,
    ) -> AbsDownloader:
        if type is None:
            if type_name is None:
                match urlsplit(url).hostname:
                    case "x.com" | "twitter.com":
                        type = PageType.twitter
                    case "www.hoyolab.com":
                        type = PageType.hoyolab
            else:
                type = PageType[type_name]
        match type:
            case PageType.twitter:
                return TwitterDownloader()
            case PageType.hoyolab:
                return HoyolabDownloader()

    def scrap(self, url):
        return self.getInstance(url).scrap(url)

File: internal/scrapper/factory.go
package scrapper

import (
	"Scrapper/internal/app"
	"Scrapper/internal/dto"
	"Scrapper/internal/entity/enum"
	"fmt"
	"net/url"
)

type factoryStruct struct {
	AbsScrapper
	Instance Scrapper
	Url      *url.URL
	Args     ScrapArgs
}

var _ Scrapper = (*factoryStruct)(nil)

func (f *factoryStruct) PreprocessURL(rawURL string) (*url.URL, error) {
	return f.Instance.PreprocessURL(rawURL)
}

func (f *factoryStruct) Scrap(args *ScrapArgs) (dto.ScrapCreate, error) {
	arg := args
	if args == nil {
		arg = &f.Args
	}
	result, err := f.Instance.Scrap(arg)

	return result, err
}

func Factory(urlStr string, pageType *enum.PageType, typeName *string, config *app.Config) (*factoryStruct, error) {
	var instance Scrapper
	var resolvedType enum.PageType

	if pageType == nil {
		if typeName == nil {
			hostname, _ := url.Parse(urlStr)
			resolvedType, _ = enum.FromHost(hostname.Hostname())
		} else {
			resolvedType, _ = enum.FromHost(*typeName)
		}
	} else {
		resolvedType = *pageType
	}
	base := AbsScrapper{resolvedType, config}
	switch resolvedType {
	case enum.Twitter:
		instance = &implTwitter{base}
	case enum.HoyoLab:
		instance = &implHoyolab{base}
	case enum.HoyoLink:
		base.PageType = enum.HoyoLab
		instance = &implHoyolink{implHoyolab{base}}
	case enum.Instagram:
		instance = &implInsta{base}
	default:
		return nil, fmt.Errorf("unsupported page type")
	}

	// URL 전처리 및 ScrapArgs 생성
	processedURL, err := instance.PreprocessURL(urlStr)
	if err != nil {
		return nil, err
	}
	args := instance.GenArgs(processedURL)

	return &factoryStruct{
		AbsScrapper: base,
		Instance:    instance,
		Url:         processedURL,
		Args:        args,
	}, nil
}
